### PR TITLE
migrations: Ensure leadership is claimed when importing a model

### PR DESCRIPTION
**Pull request by babbageclunk**
## Description of change

With raft-based leases, the import code that claims leadership for applications in the incoming model needs to move out of state - the global lease manager isn't accessible from there. We now claim leadership in `migration.Import`, passing in a function to get the leadership claimer from the API facade.

This required changing facade context.LeadershipClaimer to accept a model uuid - migrationtarget is a controller facade, so the model being imported isn't the one that we're connected to (since it doesn't exist when the connection is made).

The other wrinkle is that we can't get the legacy leadership claimer from the facade context, because we'd need to get it from the state pool, which in turn would require that we either return a LeadershipClaimer extended with a Close to release the pooled state, or a closer function. It seemed simpler just to check the feature flag and get the claimer from the state we have after importing if LegacyLeases is set. 

## QA steps

* Bootstrap 2 controllers (A and B) and deploy 5 or 6 units of an application in a model on A.
* Migrate the model from A to B.
* After the migration is finished, verify that the leader has been maintained correctly.
* In debug logging for juju.migration check that the leadership was claimed for the right unit during import (so it wasn't just chance that the leadership was carried over).

### CODE REVIEW

**Comment by babbageclunk** [comment]
Intermittent failure: https://bugs.launchpad.net/juju/+bug/1788401
Hopefully fixed by #9146 

!!build!!

File: apiserver/facades/controller/migrationtarget/migrationtarget_test.go [hunk 1]
```diff
@@ -35,9 +39,10 @@ import (
 type Suite struct {
 	statetesting.StateSuite
 	resources  *common.Resources
-	authorizer apiservertesting.FakeAuthorizer
+	authorizer *apiservertesting.FakeAuthorizer
```
Comment (wallyworld) [inline apiserver/facades/controller/migrationtarget/migrationtarget_test.go]: why this change?
when we store a facade context struct and not a pointer (for example)

File: migration/migration.go [hunk 2]
```diff
@@ -16,14 +17,23 @@ import (
...
+	"github.com/juju/juju/feature"
 	"github.com/juju/juju/resource"
 	"github.com/juju/juju/state"
 	"github.com/juju/juju/tools"
 )
 
 var logger = loggo.GetLogger("juju.migration")
 
+// When we import a new model, we need to give the leaders some time to
+// settle. We don't want to have leader switches just because we migrated an
```
Comment (wallyworld) [inline migration/migration.go]: a/an/a

File: migration/migration.go [hunk 3]
```diff
@@ -16,14 +17,23 @@ import (
...
+// model, so this time needs to be long enough to make sure we cover
+// the time taken to migration a reasonable sized model. We don't yet
```
Comment (wallyworld) [inline migration/migration.go]: s/migration/migrate

File: migration/migration.go [hunk 4]
```diff
@@ -16,14 +17,23 @@ import (
...
+// know how long this is going to be, but we need something.
+const initialLeaderClaimTime = time.Minute
```
Comment (wallyworld) [inline migration/migration.go]: Arbitrary times like this are a bit kudgy. I can see it could take longer than a minute waiting for all the agents to restart again etc. I wonder if we need a way to lock or freeze leadership claims and then unfreeze after migration.
Can we at least make it a todo to come back and take another look

File: state/upgrades.go [hunk 5]
```diff
@@ -1272,6 +1273,8 @@ func MigrateLeasesToGlobalTime(pool *StatePool) error {
 	return runForAllModelStates(pool, migrateModelLeasesToGlobalTime)
 }
 
+const initialLeaderClaimTime = time.Minute
```
Comment (wallyworld) [inline state/upgrades.go]: This appears duplicated. Can we use a single sourced value?

File: apiserver/facades/controller/migrationtarget/migrationtarget_test.go (hunk 1 above)
Comment (babbageclunk) [reply apiserver/facades/controller/migrationtarget/migrationtarget_test.go]: I needed to be able to change it in a test and have the change be visible in the constructed facade. When it wasn't a pointer then the authorizer was getting copied into the facade context, and changes made in the body of the test had no effect.

File: migration/migration.go (hunk 2 above)
Comment (babbageclunk) [reply migration/migration.go]: Oh, yup!

File: migration/migration.go (hunk 3 above)
Comment (babbageclunk) [reply migration/migration.go]: Yup - I just cut-n-pasted this without reading it closely enough, I'm blaming @mjs for the original typos.

File: state/upgrades.go (hunk 5 above)
Comment (babbageclunk) [reply state/upgrades.go]: Actually, it needed to be public in state - the other way produces an import cycle.

File: migration/migration.go (hunk 3 above)
Comment (mjs) [reply migration/migration.go]: I apologise for any migraines my migratory migrations typos may have caused :/
